### PR TITLE
Add support for AccessLog API

### PR DESCRIPTION
**Pull request by jeskepetr**
Fixes https://github.com/gooddata/gooddata-java/issues/1004

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by peter-plochan** [comment]
Please do not use GoodData internal tickets in PRs/commits for this "really public" OSS library. Create [GitHub issue](https://github.com/gooddata/gooddata-java/issues) instead and add there the description of problem/feature and possibly, if you need to, link the internal ticket within this issue.

File: gooddata-java/src/main/java/com/gooddata/sdk/service/accesslog/TimeFilterPageRequest.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import com.gooddata.sdk.common.util.GoodDataToStringBuilder;
+import com.gooddata.sdk.common.util.ISOZonedDateTime;
+import com.gooddata.sdk.common.util.MutableUri;
+import org.apache.commons.lang3.Validate;
+import org.springframework.beans.BeanUtils;
+
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+
+public class TimeFilterPageRequest extends CustomPageRequest {
```
[the remark this answers is not in the document]
Comment (peter-plochan) [reply gooddata-java/src/main/java/com/gooddata/sdk/service/accesslog/TimeFilterPageRequest.java:18]: Also, what's the purpose of this class when it is not used in any GoodData service (meaning the service within this library, usually inheritor of [AbstractService](https://github.com/gooddata/gooddata-java/blob/ac293f03758a0b9081dd1c7ef25d67090caff3ab/gooddata-java/src/main/java/com/gooddata/sdk/service/AbstractService.java))?

**Review by peter-plochan — CHANGES_REQUESTED** [review]
See comments below.
Also, I'm just wondering what's the purpose of this not-[AuditEvent](https://github.com/gooddata/gooddata-java/blob/b6f15692d3aca46c84c22c83bc7c227ff4537b2b/gooddata-java-model/src/main/java/com/gooddata/sdk/model/auditevent/AuditEvent.java#L26) bean and its special API endpoint? As I can see, it still represents a part audit log (GD internally - CFAL)...

**Comment by mverbovsky** [comment]
> See comments below.
> Also, I'm just wondering what's the purpose of this not-[AuditEvent](https://github.com/gooddata/gooddata-java/blob/b6f15692d3aca46c84c22c83bc7c227ff4537b2b/gooddata-java-model/src/main/java/com/gooddata/sdk/model/auditevent/AuditEvent.java#L26) bean and its special API endpoint? As I can see, it still represents a part audit log (GD internally - CFAL)...

It is part of cfal, but they are special events directly from haproxy that represents access logs for particular hosts. In this case you do not have domain and user context. So if you try to reuse AuditLog then you will have to change whole logic behind AuditLogs because they are build upon domain and user.

**Comment by peter-plochan** [comment]
> > See comments below.
> > Also, I'm just wondering what's the purpose of this not-[AuditEvent](https://github.com/gooddata/gooddata-java/blob/b6f15692d3aca46c84c22c83bc7c227ff4537b2b/gooddata-java-model/src/main/java/com/gooddata/sdk/model/auditevent/AuditEvent.java#L26) bean and its special API endpoint? As I can see, it still represents a part audit log (GD internally - CFAL)...
> 
> @peter-plochan It is part of cfal, but they are special events/acces logs directly from haproxy that represents access logs for particular hosts. In this case you do not have domain and user context. So if you try to reuse AuditLog then you will have to change whole logic behind AuditLogs because they are build upon domain and user. The API also will support Common Log Format, so it is really access logs to particular host.

Ahh, ok, thanks for the explanation @mverbovsky 

So the question is, if this whole thing should or should not share the Java package with the other audit log entities (`AuditEvent`).

**Comment by mverbovsky** [comment]
> > > See comments below.
> > > Also, I'm just wondering what's the purpose of this not-[AuditEvent](https://github.com/gooddata/gooddata-java/blob/b6f15692d3aca46c84c22c83bc7c227ff4537b2b/gooddata-java-model/src/main/java/com/gooddata/sdk/model/auditevent/AuditEvent.java#L26) bean and its special API endpoint? As I can see, it still represents a part audit log (GD internally - CFAL)...
> > 
> > 
> > @peter-plochan It is part of cfal, but they are special events/acces logs directly from haproxy that represents access logs for particular hosts. In this case you do not have domain and user context. So if you try to reuse AuditLog then you will have to change whole logic behind AuditLogs because they are build upon domain and user. The API also will support Common Log Format, so it is really access logs to particular host.
> 
> Ahh, ok, thanks for the explanation @mverbovsky
> 
> So the question is, if this whole thing should or should not share the Java package with the other audit log entities (`AuditEvent`).

I would vote for the same package, because it is still part of audit logging.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: gooddata-java/src/main/java/com/gooddata/sdk/service/accesslog/TimeFilterPageRequest.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    public static TimeFilterPageRequest copy(TimeFilterPageRequest source) {
+        Validate.notNull(source, "source cannot be null", new Object[0]);
+        TimeFilterPageRequest copy = new TimeFilterPageRequest();
+        BeanUtils.copyProperties(source, copy);
+        return copy;
+    }
+
+    public TimeFilterPageRequest withIncrementedLimit() {
+        TimeFilterPageRequest copy = copy(this);
+        copy.setLimit(this.getSanitizedLimit() + 1);
```
[the remark this answers is not in the document]
Comment (jeskepetr) [reply gooddata-java/src/main/java/com/gooddata/sdk/service/accesslog/TimeFilterPageRequest.java]: I am not entirely sure if it is better now when I've rewritten it. Please have a look if I've reimplemented as you've imagined.

File: gooddata-java-model/src/main/java/com/gooddata/sdk/model/auditevent/AccessLog.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.gooddata.sdk.common.util.GoodDataToStringBuilder;
+import com.gooddata.sdk.common.util.ISOZonedDateTime;
+
+import java.time.ZonedDateTime;
+
+/**
+ * Model class, used for special cfal events/access logs directly from haproxy,
```
Comment (peter-plochan) [inline gooddata-java-model/src/main/java/com/gooddata/sdk/model/auditevent/AccessLog.java:19]:
```suggestion
 * Model class, used for special audit log events/access logs directly from haproxy,
```

File: gooddata-java/src/main/java/com/gooddata/sdk/service/accesslog/TimeFilterPageRequest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (peter-plochan) [reply gooddata-java/src/main/java/com/gooddata/sdk/service/accesslog/TimeFilterPageRequest.java]: ok, you're right, I didn't realise that we need to use the `copy` method from the `AuditEventPageRequest`, so the previous one was better

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Review by peter-plochan — APPROVED** [review]
LGTM 👍

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-java-merge-compliance-check/213/ : SUCCESS in 26s
- https://checklist.intgdc.com/job/compliance/job/gooddata-java-fossa-scanning-zuul-el7/53/ : SUCCESS in 21s (non-voting)

**Pull request merged**